### ISSUE TITLE
Battle Testing

### ISSUE THREAD

**Issue by mohamedhafez** [ostinelli/apnotic#12]
Like many people probably, I'm considering moving to this gem from grocer or houston. I'm wondering if anyone has used it in production successfully and reliably, and what performance was like, and how it matches up to grocer.  (Basically, I'm hoping for a thread similar to https://github.com/alloy/lowdown/issues/11). Thanks!

**Comment by benubois** [comment]
Hi @mohamedhafez,

I switched to using apnotic shortly after I heard about it from [Ruby Weekly in April](http://rubyweekly.com/issues/295).

I had been looking to switch to something that worked with the HTTP/2 APNS since the old TCP one was tricky to work with. Previously I had used both Grocer and Houston.

I tried lowdown but didn't have any luck getting it to work. Also the dependency on celluloid was a little more heavy-duty then I was looking for as I preferred to rely on sidekiq for concurrency.

Anyway the experience with apnotic has been outstanding and rock-solid. Combined with Apple's HTTP/2 push notification service it is _so_ much better than the old way and performant enough to keep up with my use case for [Feedbin](https://feedbin.com) where I use it for and [iOS push notifications based app](https://feedbin.com/notifier) and [Safari push notifications](https://feedbin.com/blog/2013/10/24/mavericks-push-notifications/). Between these things Feedbin sends ~150k notifications/week. Not a ton, but I have no doubt it would be easy to scale it to push way more. You can [see my implementation here](https://github.com/feedbin/feedbin/blob/master/app/workers/device_push_notification_send.rb).

Performance-wise I do see a slight increase in the average time around the switchover, but it it negligible and not a bottle-neck. Also I don't consider the old APNS to be functional so it's not really a fair comparison. 

![screen shot 2016-06-14 at 10 44 27 pm](https://cloud.githubusercontent.com/assets/133809/16069520/f8b285d6-3283-11e6-80b6-7aa29441930b.png)
<img width="939" alt="screen shot 2016-06-14 at 10 59 17 pm" src="https://cloud.githubusercontent.com/assets/133809/16069521/f8b3056a-3283-11e6-8afb-e17997706bad.png">

**Comment by mohamedhafez** [comment]
@evrenios you had a really informative post in https://github.com/alloy/lowdown/issues/11#issuecomment-191869392, any chance you'd be interested in running those same tests with this gem?

**Comment by ostinelli** [comment]
@mohamedhafez we are also using it in production since a couple of months. 

@adamvduke may also have additional feedback: care to share? :)

**Comment by benubois** [comment]
Quick update. After updating to v0.9.2 the performance increased and it is now faster than grocer 🎉 

<img width="926" alt="screen shot 2016-06-29 at 6 04 45 pm" src="https://cloud.githubusercontent.com/assets/133809/16473792/74304624-3e25-11e6-9baf-958ea4311bb6.png">

**Comment by ostinelli** [comment]
Yes, this is due to an optimization asked in https://github.com/ostinelli/net-http2/issues/6. Glad to see that it's working for you too.

**Comment by edonosotti** [comment]
@benubois: this is slightly off-topic, but can I ask you what metrics tracking tool did you use to measure performances and plot those graphs? Thank you!

**Comment by benubois** [comment]
@edonosotti that's [Librato](http://librato.com). Highly recommend it.

**Comment by ebalassanian** [comment]
I have the gem fully integrated into a project that heavily uses notifications to manage distributed client state. It's worked perfectly. I do not have metrics to share yet, but will publish them here when ready.

**Comment by ostinelli** [comment]
@ebalassanian sounds good.

BTW: I recommend to anyone using in production to ensure that you use the latest Apnotic `0.10.2` with Net-Http2 `0.13.3`, they have considerable increased stability.

**Comment by ostinelli** [comment]
FYI: Apnotic has been used in production for months now. It has successfully sent millions of notifications, hence I've just released `1.0.0`. Any feedback is welcome :)

**Comment by elhu** [comment]
👍

**Comment by j0sh** [comment]
Doing some light testing with this gem and one thing I'm seeing is that after a certain length of time, connections to APNS time out, regardless of what the `timeout` parameter is set to. (I think the default timeout is 30s, while it takes ~15 minutes to time out here. Here is what the backtrace fails with: [2].

Are there any best practices for keeping the connection alive over a sporadically used connection? And if the connection does fail eventually, will the HTTP2 client take care of reconnects in pooled mode, as the ConnectionPool library says it should [1]? Testing this scenario right now but figured I'd ask here as well. Using dev servers, for what it's worth.

[1] https://github.com/mperham/connection_pool
[2] https://gist.github.com/j0sh/a215eebb8f1f0cc4d6f6270bc02df352

**Comment by ostinelli** [comment]
Hi @j0sh, these are issues that don't belong in here. Please open a new issue where we can discuss this.

**Comment by ostinelli** [comment]
@ebalassanian any data you'd like to share? I'm currently seeing if any optimizations are necessary.

**Comment by eebs** [comment]
Just wanted to chime in and say thanks for all the work you and others have put in @ostinelli. We've been using Apnotic in production for 15 months, and have sent 6.8 million push notifications.

After working through some threading issues ([the addition of the `:error` callback helped tremendously](https://github.com/ostinelli/apnotic/issues/48#issuecomment-306426901)), things are quite stable.

Thanks again!

**Comment by ostinelli** [comment]
Thank you @eebs, glad Apnotic is doing its part! :)

**Comment by bankair** [comment]
Being using apnotic (mostly MDM notifications) since a couple of months. So far, so good. We are migrating our old notification engine from apn_sender to apnotic right now.

I don't have clear KPIs as of today, but we did not encounter unexplained issue.

**Comment by benrugg** [comment]
This is an old thread, but I wanted to chime in on it. I have had to change the entire push notification architecture for my app three times over the years as different libraries stopped being maintained, so before I switched to apnotic, I really wanted to make sure it was going to be solid.

My implementation is with apnotic running on sidekiq, sending 150k-200k notifications a day. I've had it running for two weeks, and it's been rock solid.

![sidekiq report](https://user-images.githubusercontent.com/1221274/49103796-eabf4600-f241-11e8-9036-1d3daf1d0208.png)

Most recently before this, I was using [rpush](https://github.com/rpush/rpush), and having numerous issues that affected reliability. Some were architectural problems (notifications getting stuck in a pending state because of failed notifications earlier in their batch), but I was also seeing signifcant memory leaks, and I had four separate times where rpush (running as a daemon) hung and stopped sending any notifications until I restarted it!

Apnotic / sidekiq / redis have been perfectly stable so far, and are using orders of magnitude less memory.

My top speed sending notifications has been about 800 / second. I have moments during the day where 5k-7k notifications need to go out at the same time, and this takes 10-14 seconds. This is by far the best performance I've had, but it's also lower than I would have expected. (I have 5 connections to APNs and 10 sidekiq threads, running on an m5.large EC2 instance).

TL;DR: Apnotic is very stable and more performant than any other library I've used.

**Comment by ostinelli** [comment]
Thank you for your feedback!
